### PR TITLE
[Workspace] Dev branch merge attempt

### PR DESCRIPTION
**Pull request by jlamypoirier**
# ✨ Description

Workspace for dealing with the merge. Not intended to work yet.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -29,16 +29,17 @@ ENV PIP_CONSTRAINT=""
 # There is no pre-build mamba image for pytorch 2.8, we build it before the rest to avoid rebuilds.
 # We need to compile from the repo because of https://github.com/state-spaces/mamba/issues/720 (same for causal-conv1d)
 # We set the number of workers to avoid OOM when compiling on laptop. (TODO: Can we make it configurable?)
+# Using varlen_mamba for variable length sequence support
 RUN MAX_JOBS=2 pip install --no-build-isolation  "causal-conv1d@git+https://github.com/Dao-AILab/causal-conv1d@2a288a1"
-RUN MAX_JOBS=2 pip install --no-build-isolation "mamba_ssm[causal-conv1d]@git+https://github.com/state-spaces/mamba@4a8a2a2"
+RUN MAX_JOBS=2 pip install --no-build-isolation "mamba_ssm[causal-conv1d]@git+https://github.com/jxiw/varlen_mamba@varlen_mamba"
```
Comment (jlamypoirier) [inline Dockerfile:34]: Ignoring varlen mamba for now

File: fast_llm/functional/config.py [hunk 2]
```diff
@@ -67,7 +68,8 @@ def _set_activation_fn_map() -> None:
     global _ACTIVATION_FN_MAP
 
     _ACTIVATION_FN_MAP = {
-        ActivationType.gelu: lambda x: torch.nn.functional.gelu(x, approximate="tanh"),
+        ActivationType.gelu: torch.nn.functional.gelu,
```
Comment (jlamypoirier) [inline fast_llm/functional/config.py:71]: We can't change this

File: fast_llm/layers/vision_encoder/adapter.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
+import typing
```
Comment (jlamypoirier) [inline fast_llm/layers/vision_encoder/adapter.py:1]: This is just a MLP, let's make it one.

File: fast_llm/layers/vision_encoder/config.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+    kv_channels = "vit_kv_channels"
+    max_image_tokens = "max_image_tokens"
+    patch_embeddings = "patch_embeddings"
+    hidden_dims = "vit_hidden_dims"
+    image_patches_meta = "vit_image_patches_meta"
+    out_channels = "vit_out_channels"
+
+
+@config_class()
+class ImageNormalizationConfig(Config):
```
Comment (jlamypoirier) [inline fast_llm/layers/vision_encoder/config.py:41]: Belongs to dataset preprocessing

File: fast_llm/layers/vision_encoder/preprocessing.py [hunk 5]
```diff
@@ -0,0 +1,281 @@
+import math
```
Comment (jlamypoirier) [inline fast_llm/layers/vision_encoder/preprocessing.py:1]: Moving most of this to dataset preprocessing

File: fast_llm/data/dataset/gpt/sampled.py [hunk 6]
```diff
@@ -468,21 +531,97 @@ def __getitem__(self, index: int) -> typing.Any:
...
-                token_end_index_in_document = min(token_end - token_count, document_size)
+                token_end_index_in_document = min(token_end - token_count, text_size)
                 sample = self._indexed_dataset.get(
                     document_index,
                     offset=token_start_index_in_document,
                     length=token_end_index_in_document - token_start_index_in_document,
                     use_loss_masking_spans=self._parameters.use_loss_masking_spans,
                 )
-                token_ids.append(sample.token_ids)
+                start_pos = 0
```
Comment (jlamypoirier) [inline fast_llm/data/dataset/gpt/sampled.py:552]: All this mess is there so we know how many tokens the images will take, because preprocessing is done in the model. Moving to dataset preprocessing (before sampling) to simplify.

File: fast_llm/models/gpt/model.py [hunk 7]
```diff
@@ -67,8 +146,41 @@ def preprocess_meta(
             micro_sequence_length = sequence_length
             truncate_documents = True
 
-        batch_data = self._distributed_config.get_distributed_dim(DistributedDimNames.batch_data)
-        batch_dim = TensorDim(BlockDimNames.batch, micro_batch_size * batch_data.size, batch_data)
+        if self._config.vision_encoder.enabled:
```
Comment (jlamypoirier) [inline fast_llm/models/gpt/model.py:149]: Unnecessary complexity, not needed once we move preprocessing

File: fast_llm/data/dataset/gpt/memmap.py [hunk 8]
```diff
@@ -26,32 +28,46 @@ def __init__(
...
         self._prefix = pathlib.Path(prefix)
         self._has_spans = 0
+        self._has_images = 0
         self._has_preference_spans = False
 
         with self._prefix.with_suffix(".idx").open("rb") as stream:
             Assert.eq(stream.read(9), MEMMAP_INDEX_HEADER, msg=f"File: {stream.name}")
             self._version = struct.unpack("<Q", stream.read(8))[0]
-            assert self._version in [1, 2, 3], f"Unsupported version for gpt_memmap dataset: {self._version}."
+            assert self._version in [1, 2, 3, 4], f"Unsupported version for gpt_memmap dataset: {self._version}."
```
Comment (jlamypoirier) [inline fast_llm/data/dataset/gpt/memmap.py:53]: This is not sustainable. Switching to a json header.

File: fast_llm/data/dataset/gpt/memmap.py [hunk 9]
```diff
@@ -120,16 +135,53 @@ def _init(self, name: str, prefix: pathlib.Path | str, num_documents: int | None
                         offset=rejected_span_offset + idx * 2 * np.dtype(np.int32).itemsize,
                     )
                 )
+            offset += np.array(self._chosen_spans).nbytes + np.array(self._rejected_spans).nbytes
```
Comment (jlamypoirier) [inline fast_llm/data/dataset/gpt/memmap.py:138]: This is inefficient. We can just store the image count cumsums instead. (Same for spans BTW)

File: fast_llm/functional/cross_entropy.py [hunk 10]
```diff
@@ -151,7 +164,8 @@ def _fused_cross_entropy_forward_backward(
 
     loss = per_sample_loss.mean()
     if target_format != TargetFormat.labels and group is not None:
-        all_reduce(loss, op=ReduceOp.MEAN, group=group)
+        all_reduce(loss, op=ReduceOp.SUM, group=group)
```
Comment (jlamypoirier) [inline fast_llm/functional/cross_entropy.py:167]: Why?

File: fast_llm/layers/language_model/head.py [hunk 11]
```diff
@@ -11,7 +11,13 @@
 from fast_llm.engine.config_utils.tensor_dim import TensorDim, scalar_dim
 from fast_llm.engine.distributed.config import DistributedConfig, DistributedDimNames
 from fast_llm.functional.autograd import grad_is_context, wrap_forward_backward
-from fast_llm.functional.config import CrossEntropyImpl, DistillationLossImpl, TargetFormat, TritonConfig
+from fast_llm.functional.config import (
```
Comment (jlamypoirier) [inline fast_llm/layers/language_model/head.py:14]: Ignoring changes to model head and reverse kl for now

File: fast_llm/engine/multi_stage/stage.py [hunk 12]
```diff
@@ -153,7 +153,7 @@ def backward(
         assert self._mode.support_backward
         input_, output = grad_context
         output.backward(output_grad)
-        return input_.grad
+        return input_.grad if input_.grad is not None else torch.zeros_like(input_)
```
Comment (jlamypoirier) [inline fast_llm/engine/multi_stage/stage.py:156]: Bad idea. This adds overhead to the first layer for all models

File: tests/layers/test_lm_head.py [hunk 13]
```diff
@@ -23,12 +23,10 @@ def _reverse_kl_loss(
 ):
     scaled_target = target / teacher_softmax_temperature
 
-    scaled_target = torch.clamp(target, min=-50, max=50)
```
Comment (jlamypoirier) [inline tests/layers/test_lm_head.py:26]: Why?

File: fast_llm/layers/multi_modal/embedding.py [hunk 14]
```diff
@@ -0,0 +1,183 @@
+import typing
```
Comment (jlamypoirier) [inline fast_llm/layers/multi_modal/embedding.py:1]: Much simpler solution that can go directly in `LanguageModelEmbedding` and generalizes to other multimodal models:
* Move token ids to kwargs, replace input by the image or other embeddings. Same as here, but do it in the base class too. Use a placeholder or `None` for the LLM non-existant input.
* Use a pre-built map (tensors) from input_ (image embeddings) to LM embeddings, and copy image/multimodal embeddings with a one-liner.

File: fast_llm/layers/multi_modal/embedding.py [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+                                sample_idx, row_start_src : row_start_src + patch_width
+                            ]
+                    else:
+                        embeddings[sample_idx, position : position + num_patches] = input_[
+                            sample_idx, image_embedding_offset : image_embedding_offset + num_patches
+                        ]
+                    # Move to the next image in the input tensor
+                    image_embedding_offset += num_patches
+
+            if self._use_absolute_position_embeddings:
```
Comment (jlamypoirier) [inline fast_llm/layers/multi_modal/embedding.py:155]: Position embeddings should be masked?

File: fast_llm/layers/multi_modal/embedding.py [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+                embeddings = embeddings + torch.nn.functional.embedding(position_ids, self.position_embeddings_weight)
+            if self._sequence_parallel:
+                embeddings = split(embeddings, group=group, dim=0)
+        else:
+            if self._sequence_parallel:
+                tokens = split(tokens, group=group, dim=0)
+                if self._use_absolute_position_embeddings:
+                    position_ids = split(position_ids, group=group, dim=0)
+            # mask padded tokens
+            token_mask = tokens >= 0
```
Comment (jlamypoirier) [inline fast_llm/layers/multi_modal/embedding.py:129]: Is there a one-to-one correspondence between masked tokens and those replaced by image embeddings? If so this seems a bit redundant, and there are better solutions...

File: fast_llm/layers/multi_modal/embedding.py [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+                        embeddings[sample_idx, embedding_start_index:embedding_end_index] = input_[
+                            input_start_index:input_end_index, sample_idx
+                        ]
+                        # embeddings[sample_idx, position : position + num_patches] = input_[
+                        #     sample_idx, image_embedding_offset : image_embedding_offset + num_patches
+                        # ]
+                    image_embedding_offset += num_patches
+                    if image_embedding_offset > patch_end_offset:
+                        break
+            embeddings = reduce_forward(embeddings, group)
```
Comment (jlamypoirier) [inline fast_llm/layers/multi_modal/embedding.py:118]: I think this is very wrong. The image embeddings are not vocab-parallel, so this incorrectly multiplies the image embeddings by TP.

Comment (jlamypoirier) [reply fast_llm/layers/multi_modal/embedding.py:118]: Actually this might be correct in a really complicated way that mixes the gathering of the sequence-parallel vision encoder outputs with their mapping to the lm embeddings. Not sure about non-sequence-parallel. Either way, this needs simplification.

File: fast_llm/layers/vision_encoder/preprocessing.py [hunk 18]
```diff
@@ -0,0 +1,281 @@
...
+                        torch.full((padding_size,), 0).to(device=self._tensor_space.distributed.device),
+                    ]
+                )
+            )
+            assert patches[-1].size(0) == kwargs[TransformerKwargs.sequence_length]
+        patches = torch.cat(patches)
+        patch_position_ids = torch.cat(patch_position_ids)
+        kwargs[VisionEncoderKwargs.image_patches] = patches
+        kwargs[VisionTransformerKwargs.patch_position_ids] = patch_position_ids
+        kwargs[VisionEncoderKwargs.rotary_inv_freq] = create_inv_freqs(
```
Comment (jlamypoirier) [inline fast_llm/layers/vision_encoder/preprocessing.py:250]: Unused?

File: fast_llm/layers/multi_modal/embedding.py (hunk 14 above)
Comment (tscholak) [reply fast_llm/layers/multi_modal/embedding.py:1]: These are all inputs, it's a bit limiting to have to declare one the "official" input and make the others kwargs. Btw, the generic case is multiple embeddings (images, audio, etc) and token id maps not just one

File: fast_llm/layers/language_model/head.py (hunk 11 above)
Comment (tscholak) [reply fast_llm/layers/language_model/head.py:14]: ok. cc @oleksost

File: fast_llm/functional/config.py (hunk 2 above)
Comment (jlamypoirier) [reply fast_llm/functional/config.py:71]: I think that's possible, but is it absolutely needed? The two are usually safe to swap, so if it's just to convert HF models I'd rather just convert everything to tanh gelu.

File: fast_llm/layers/vision_encoder/adapter.py (hunk 3 above)
Comment (jlamypoirier) [reply fast_llm/layers/vision_encoder/adapter.py:1]: Main benefit is we get the implementation for free, no new code needed. Flexibility is just a side effect.

File: fast_llm/layers/multi_modal/embedding.py (hunk 14 above)
Comment (jlamypoirier) [reply fast_llm/layers/multi_modal/embedding.py:1]: Having a single input is a restriction from the engine, so not much we can do ATM. The new version (#369) is generic in the sense that it copies any kinds of embeddings from previous layers into token ids regardless of their source. It's still not combining inputs if they come from multiple previous layers, but that will be relatively straightforward to do when we need it.
